src/sql/workbench/services/notebook/browser/models/cell.ts

```diff
@@ -336,19 +340,42 @@ export class CellModel implements ICellModel {
 
 					// requestExecute expects a string for the code parameter
 					content = Array.isArray(content) ? content.join('') : content;
-					const future = kernel.requestExecute({
-						code: content,
-						stop_on_error: true
-					}, false);
-					this.setFuture(future as FutureInternal);
-					this.fireExecutionStateChanged();
-					// For now, await future completion. Later we should just track and handle cancellation based on model notifications
-					let result: nb.IExecuteReplyMsg = <nb.IExecuteReplyMsg><any>await future.done;
-					if (result && result.content) {
-						this.executionCount = result.content.execution_count;
-						if (result.content.status !== 'ok') {
-							// TODO track error state
-							return false;
+					if (tryMatchCellMagic(this.source[0]) !== ADS_EXECUTE_COMMAND) {
+						const future = kernel.requestExecute({
+							code: content,
+							stop_on_error: true
+						}, false);
+						this.setFuture(future as FutureInternal);
+						this.fireExecutionStateChanged();
+						// For now, await future completion. Later we should just track and handle cancellation based on model notifications
+						let result: nb.IExecuteReplyMsg = <nb.IExecuteReplyMsg><any>await future.done;
+						if (result && result.content) {
+							this.executionCount = result.content.execution_count;
+							if (result.content.status !== 'ok') {
+								// TODO track error state
+								return false;
+							}
+						}
+					} else {
+						let result = extractCellMagicCommandPlusArgs(this._source[0], ADS_EXECUTE_COMMAND);
+						// Similar to the markdown renderer, we should not allow downloadResource here
+						if (result?.commandId !== '_workbench.downloadResource') {
```

Comment:
How concerned are we with arbitrary commands being executed? I wonder if we should prompt the user that the notebook is trying to execute a command (with settings to disable asking for this notebook/all notebooks).

Reply:
Notebooks are already executing arbitrary code, but I think this is a good point that I've been going back and forth about. Originally, I had this functionality turned off by default and only accessible by turning on a setting, but there's no discoverability there, so I don't love that solution.

I can look into a quick pick here with the options that you described if you think that's a preferable. At least it solves some of the discoverability problem.